### PR TITLE
Refactor middleware

### PR DESCRIPTION
**Pull request by florimondmanca**
Prompted by https://github.com/encode/httpx/pull/305#issuecomment-529144881

Moves the existing middleware in `middleware.py` into their own module in a `middleware` package.

Next up, a basic auth refactor including:
- Add and use `to_bytes()`
- Extract a reusable `build_basic_auth_header()` helper.

### CODE REVIEW

**Comment by sethmlarson** [comment]
Could we do those extras in this PR too under the cover of "refactor middlewares"?

**Comment by florimondmanca** [comment]
> Could we do those extras in this PR too under the cover of "refactor middlewares"?

Uh, I guess I'm OK with that. I have another idea for refactoring the `RedirectMiddleware` to use contextvars, but that will *definitely* be in another PR. ;-) I'll push the extras.

**Comment by sethmlarson** [comment]
Totally agree. You might have some minor conflicts over the PR I just merged.

**Comment by florimondmanca** [comment]
Already force-pushed a rebase. :+1:

File: httpx/utils.py [hunk 1]
```diff
@@ -169,3 +169,7 @@ def get_logger(name: str) -> logging.Logger:
             logger.addHandler(handler)
 
     return logging.getLogger(name)
+
+
+def to_bytes(value: typing.Union[str, bytes]) -> bytes:
```
Comment (sethmlarson) [inline httpx/utils.py]: We should take an encoding parameter here too, default to UTF-8.

Comment (florimondmanca) [reply httpx/utils.py]: Fixed!

**Pull request merged**